### DISCUSSION TITLE
About the Live Streaming Issue with RTMP Chunk Buffer, Custom Video Effect when RTMP Stream and Stream Recorder simultaneously run

### DISCUSSION THREAD

**Discussion by wchklaus97** [HaishinKit/HaishinKit.swift#1608] [Q&A] [answered]
**Version of 'HaishinKit', '2.0.0-rc.1'**

I am working on an R&D project about complex layers in live streaming with simultaneous saving of the live stream into photos on iOS. The project involves frequent updates to the complex layer, up to every 0.1 seconds.

I noticed your last commit (697c79390d0ce1673519c40ad30e2f115092f21b) where you improved and optimized the usage of RTMPChunkBuffer handling. I am facing similar issues and wanted to ask about the kinds of problems you encountered.

My project is causing issues due to heavy video effects being applied with frequent updates (as I have to convert custom UI in Swift to CIImage and trigger video effect updates). I found that there's a 20% chance of encountering ChunkBuffer issues in my project. Additionally, I've noticed that under these complex conditions, CPU and memory usage grow very rapidly (I have optimized thread safety and queue execution in my project).

However, I haven't yet applied any solutions to handle the reuse of the complex layer. Before optimizing my project further, I would like to hear your ideas. Thank you.

The screenshot of the RTMPChunkBuffer issue is attached below.

![Screenshot 2024-10-20 at 2 08 33 PM](https://github.com/user-attachments/assets/d2c59750-f88b-4582-8447-8d2afe462cb3)

![Screenshot 2024-10-20 at 2 08 40 PM](https://github.com/user-attachments/assets/fcd7cdf5-3950-4f71-b860-0d0f9ef77f53)

![Screenshot 2024-10-20 at 2 08 48 PM](https://github.com/user-attachments/assets/d9031ed8-b7c6-471b-9d35-ea3db1dd123f)

![Screenshot 2024-10-20 at 2 08 57 PM](https://github.com/user-attachments/assets/a3800f16-6e01-4a28-9a97-95febb6ed294)

![Screenshot 2024-10-20 at 2 09 06 PM](https://github.com/user-attachments/assets/9f6c247d-01c4-4b83-bde4-5fea19cd5c73)

![Screenshot 2024-10-20 at 2 09 11 PM](https://github.com/user-attachments/assets/4fcb3ccd-73c0-42af-a2cd-58e582f18272)

In my current setup, I have applied two queues: one called updateQueue used to handle updates, and another called renderQueue used to handle rendering. This approach was implemented to avoid issues with the 'RTMP Chunk Buffer'. I found that while this solution works, it is only suitable for lightweight custom layers (simple UI with fewer video effects).

**Comment by shogo4405** [comment, marked as answer]
Simply put, crashes related to buffer overrun/underrun were occurring. Memory optimization has been implemented, and memory is now properly allocated to prevent buffer overruns. https://github.com/shogo4405/HaishinKit.swift/discussions/1605

The issue was occurring when the video bitrate was increased to 2Mbps

**Reply by wchklaus97** [reply]
Sorry to bother you, but is the buffer handling feature also ready to be added and supported in version 2.0.0-rc.1? Since I am using this version for my R&D, it would be great if the buffer handling feature could be included in this pre-release version, as it would greatly assist me in my further research and development.

**Reply by shogo4405** [reply]
No, it’s not possible since version 2.0.0-rc.1 has already been released. It will be included in the next 2.0.0-rc.2 release.

Since you are in the research stage, I thought you could pull from the `main branch`, but would that be too much trouble? If you are using SPM or CocoaPods, I believe you should be able to do it.
